### ISSUE TITLE
Unable to add layout (background) plugin

### ISSUE THREAD

**Issue by PeterKottas** [react-page/react-page#752]
I am unable to drop in the background layout plugin (all content plugins can be added no problem). I wanted to run this past you @macrozone, maybe you can reproduce easily. I'll do more digging if it works for you fine but I think you might have changed something around layout plugins that might be causing this.

**Comment by PeterKottas** [comment]
I initially suspected react-dnd but CELL_INSERT_BELOW is fired properly. Also the action payload makes sense from what I see (it contains the correct item and the plugin field there is configured properly)

**Comment by PeterKottas** [comment]
It's just not added to the tree, causes the editor to rerender though and onChange is fired as well. However, the new plugin is not there.

**Comment by macrozone** [comment]
@PeterKottas its working on the example, i just verified it. But i observed a bug where it would sometimes not add a plugin when its the last in the editor, not sure if its related to. Do you see any error in the console?

i don't know if there are some changes that would break it... 🤔

can you check your yarn or package lock to check if all plugins are on the same version?

**Comment by PeterKottas** [comment]
Crap, that's what I was afraid of ... I see this in the console: 
```
Material-UI: you cannot use a custom insertionPoint and <StylesContext injectFirst> at the same time.
```
But IMHO that's unrelated. (We should fix that too at some point though).
The rest of the setup is pretty straight forward. Not sure what the problem is. I am using the Editable component, not the new editor.

**Comment by macrozone** [comment]
@PeterKottas can you try to use an older version to see when the bug was introduced?

is it the only plugin that has this problem?

**Comment by macrozone** [comment]
@PeterKottas can you check.

- does the react-devtools show the properties of that plugin in the cell that should render the background?
- does it show in the dom inspector but is just empty?
- do you define `createInitialChildren` on the plugin? 

maybe it will be created with empty content and is therefore removed (optimized), i don't know the details, but i saw that there is some algorithm like that in place

**Comment by PeterKottas** [comment]
That might be tricky cause I had to make a lot of changes locally to support the 0.16 branch. I went back to beta 5 but still the same problem. Yes it's only the background plugin. Although that's the only layout plugin I have. So it might be all layout plugins. I'll try to add one extra to see if that helps.

I switched to default background plugin provided by the lib. So that should work right of the box.

**Comment by PeterKottas** [comment]
Yep, plugin is there on the component, also the initial children seem to work.

**Comment by PeterKottas** [comment]
And also all version are correctly pointed at 0.16.0-beta.9 just did a clean install as well. The plugin is not there at all (inspecting DOM). Very weird.

**Comment by macrozone** [comment]
Is it a custom background plugin? can you show the source code? It will be a bit tricky without a reproduction. maybe you can integrate your plugin into the example?

maybe you can debug the rendering of the cell? So the prop is there on the data, maybe you can find out why it does not get rendered. Maybe it has to do with the draft property?


(PS: it should still work with the "old" approach with separated editable and ui, the new package just wraps both into one and handles sticky placement of the buttons on the right. )

(also sorry for no clean branches, i should start to do more atomic commits for easier bug tracking)

**Comment by PeterKottas** [comment]
nah, it's just standard bg plugin:
```
import background from '@react-page/plugins-background';
let bgPlugin = background({
  imageUpload: undefined,
  defaultPlugin: slatePlugin,
  // Renderer: imageBackground,
});
```
Nothing fancy, I also copy pasted the customLayoutPlugin from examples. Same problem. So it's a problem with layout plugins in general.

Yeah I prefer the new approach actually, but can't change everything right away. That being said, it should definitely work. Np, now is a turbulent time. Lot of new features. It takes time to clean up.

**Comment by macrozone** [comment]
nothing special there. is the slatePlugin the default slatePlugin or a customized one?

do you pass the layout plugins properly to the editor (and html renderer) in `layout` ?

**Comment by PeterKottas** [comment]
just standard slate
```
import slate from '@react-page/plugins-slate';
let slatePlugin = slate();
````
the these are all plugins
```
const oryPlugins: Plugins = {
  content: [slatePlugin, spacer, imgPlugin, video, divider, html5video],
  layout: [bgPlugin, customLayoutPlugin()],
  native,
};
```
Editor
```
const oryEditor = new Editor({
  plugins: oryPlugins,
  defaultPlugin: slatePlugin,
  editables: [emptyEditorState],
  // tslint:disable-next-line: no-any
  store: window.store as any,
});
```

And finally the Editable
```
<Editable
  editor={oryEditor}
  id={this.props.editorState.id}
  onChange={editorState => {
    this.onEditorStateChange(editorState);
  }}
/>
```

**Comment by macrozone** [comment]
probably not related, but can you remove the `store` property? react-page now  uses its own context for redux, so that it does not "leak". Every instance of the editor has its own redux store.

**Comment by PeterKottas** [comment]
Removed, no effect. Isn't it strange that content plugins work but layout doesn't? ... I find it really odd.

**Comment by macrozone** [comment]
@PeterKottas can you try to set a breakpoint here: https://github.com/react-page/react-page/blob/master/packages/core/src/components/Cell/Inner/index.tsx#L57 ?

There are two conditions there that would prevent your cell from rendering

**Comment by PeterKottas** [comment]
It never goes into that if. In other words, LayoutComponent is never defined. Although that would happen if the plugin is never injected into the state. Which I think it is.

**Comment by PeterKottas** [comment]
Ok when I switch to the new editor approach. It works, but obviously I lose all of my custom Ui. I need to investigate more and compare my code to what is in the new editor package.

**Comment by macrozone** [comment]
@PeterKottas ok, that narrows it down.... let me check my changes

**Comment by PeterKottas** [comment]
I have a feeling it will be context related. Somehow maybe when I drop the plugin, it cannot find the correct redux store to use? Something along those lines.

**Comment by PeterKottas** [comment]
Damn ... fixed it. So no idea what it is, but apparently this is new:
```
const children = sanitizeInitialChildren(
  plugin.createInitialChildren()
);
```
I mean the sanitizeInitialChildren, this must have been added at some point and that broke my custom code. All good now.

**Comment by macrozone** [comment]
@PeterKottas haha, so i was right : https://github.com/react-page/react-page/issues/752#issuecomment-557818566 ;-)

createInitialChildren should work as before, it just supports a new notation:

```
createInitialChildren: () => {
    return [
      [
        {
          content: {
            plugin: slate,
            state: slate.createInitialSlateState(({ plugins }) => {
              return {
                children: [
                  {
                    plugin: plugins.headings.h2,
                    children: ["A title"],
                  },
                  {
                    plugin: plugins.paragraphs.paragraph,
                    children: [
                      "Some text blabla... ",
                    ],
                  },
                ],
              };
            }),
          },
        },
      ],
    ];
````


sanitizeInitialChildren is internally used to support this notation, when it receives an array. But it seems i messed it up, can you show what you returned in `createInitialChildren` (i guess an array)

**Comment by PeterKottas** [comment]
Yep this is the code from the bg-plugin
```
createInitialChildren:
      settings.getInitialChildren ||
      (() => {
        return [[{ content: { plugin: settings.defaultPlugin } }]];
      }),
```
So it is indeed an array. Ok now I see the point of that change. Basically it hides away the need of generating unique ids. Not sure if there's anything that can be done to make this more straight forward. It's not hard to grasp, but when you don't know it's needed, you can easily forget. Maybe some warning somewhere saying the children type is incorrect?

**Comment by macrozone** [comment]
@PeterKottas this is the new version. Can you show me what you had in place i your background-plugin and how you fixed the issue?

**Comment by PeterKottas** [comment]
I didn't touch the plugin, the problem was in my custom Toolbar. I've created that based on code from some time ago. Before the sanitize thing existed. Now I had to add it and it started behaving properly. Like I am saying, easy fix when you know what's missing. Tbh, this will only be problem for a handful of people who created custom UI and then want to upgrade. Hopefully they will find this issue.
